### DISCUSSION TITLE
How to recursively download system dependencies with qlot?

### DISCUSSION THREAD

**Discussion by daninus14** [fukamachi/qlot#302] [Q&A]
I have a few projects for which I'm using qlot because the dependencies are not in quicklisp.

Lets say I have the following project dependencies:

```
P2 -> P1
P3 -> P2
```

Now in in `P2` I do `qlot add P1`, and in `P3` I do `qlot add P2`.

How can I do so that when I do `qlot install` in `P3` it will get the dependencies of its dependencies, namely `P1`, without me having to declare that dependency in `P3` because it can read it from the `qlfile` in `P2`.

Is there a mistake I made in how I'm managing the project? Or is there an option I'm missing from `qlot install`?

**Comment by daninus14** [comment]
Note that quicklisp does this automatically. I think doing this would make qlot be very helpful for replacing quicklisp altogether.

See this discussion: https://www.reddit.com/r/lisp/comments/13vpa5r/common_lisp_packageproject_manager_for/ 

it seems it's a fairly common issue.

**Comment by daninus14** [comment]
I'm thinking of implementing this and submitting a PR. 

This is what it would look like:

1. Download all systems with `qlot install` (current functionality)
2. Then check for each system downloaded if it has a `qlfile`
3. If it does have a `qlfile` then proceed to download those systems as well
4. Go back to step `2` for each new system downloaded until no more `qlfile`s found.

The tricky part is conflict resolution: when two systems have the same system dependency.
- If one system has a more specific version than the other, choose that one. For example, one system simply lists in its `qlfile` a system dependency from github, whereas the other actually tags a specific branch or commit
- If both systems have specific versions which conflict with one another, then: 
    - signal a recoverable error for user input and give the user the following options:
        - Abort
        - Display System A's Dependency System's version (meaning which system declares this dependency and which version)
        - Display System B's Dependency System's version
    - If the user picked a specific version, then: 
        - that version of the system is saved into his root `qlfile`, that way any `qlot install`s in the future will already have the conflict resolved.
        - The version is marked with an `:overwrite` key with a list of values `((:source github.com :git-hash awdin12uheb23uhe23un))` in order to mark which conflicts should this dependency overwrite. This is useful in case the user adds another system in the future with a specific version of a dependency which conflicts with his dependency, so this will signal the conflict to the user and give him the choice to decide how to deal with the new conflict. This is very helpful for finding problems when installing instead of when things break later on.

So for example, the `qlfile` system could then have something like this:

```common-lisp
github fukamachi/sxql :branch some-branch :overwrite ((:source "github.com" :git-hash "awdin12uheb23uhe23un") (:source "quicklisp" :release "sxql-20231021-git"))
```

I think this basically takes care of most cases and makes it so that the conflicts only have to be resolved once, and are clearly marked for the future.

@fukamachi What do you think? Would you be open for a PR or collaborating on this?

**Reply by fukamachi** [reply]
Looks great.
As you mentioned, conflict resolution should be tricky, which is why I'm not positive about doing it.
It's not easy to determine which one is newer since there's no incremental version in Common Lisp libraries, unlike other languages.
I think most of the problems are covered in your idea.
I'm still not sure if adding `:overwrite` key to qlfile is the best, though.

**Reply by daninus14** [reply]
> It's not easy to determine which one is newer since there's no incremental version in Common Lisp libraries, unlike other languages.

Right. I was also worried about cases of systems coming from different sources like two github repositories. That would make it unclear which system to pick, so that's why the strategy to just let the user decide. The idea is basically to avoid any hairy decisions and just let the user decide, yet at the same time record that decision so that it doesn't have to be made again on every install.

> I'm still not sure if adding :overwrite key to qlfile is the best, though.

Do you think it makes more sense to add it to `qlfile.lock` or not at all? I just wanted to deal with the case where the user already gave his input and installs again, not to fail. This is particularly useful when deploying code in a server and saving exact state. Yet at the same time, if a conflict pops up again from a different source with the same system, this avoids making assumptions and lets the user decide what he wants to do. That's why I proposed saving which system source and version or git hash got overwritten.

`qlfile.lock` is actually just s-expressions, which probably make it much easier to deal with in terms of implementing it. I guess it's also stored in the repository, so it would maintain the information. I just thought that for a user forced update of systems (AKA deleting the `qlfile.lock` file and doing another `install`) it would not keep that decision in the record. Since the `qlfile` is where all the user input is actually stored and not "computed input", I thought it would be appropriate.

What are your thoughts on this? I defer to you on how to implement it.

**Reply by fukamachi** [reply]
> Do you think it makes more sense to add it to qlfile.lock or not at all?

Yeah, I totally agree with including the information in repositories; however, at the same time, I'd like to make qlfile clean as much as possible so that users to edit the file by hand.
I think `qlfile.lock` is the right place to put it since it should contain all the information needed to reproduce the same set of libraries in different environments.

Or, I don't think it's necessary at the first version while I love incremental development.
